File: src/screens/VehicleScreen.js
import React, { useState } from "react";
import { View, Text, ScrollView, TextInput, Alert } from "react-native";
import { Card, CardRow, SectionHeader, Label, useC } from "../components/ui";
import { TYPE, SPACE } from "../theme";
import { TYPE_META, getStatus, attentionLabel, getServicePackages } from "../data/templates";
import { formatDate, formatOdo, formatLogLine } from "../format";
import { exportPdf } from "../data/export";
import TopBar from "../components/TopBar";
import { PhotoStrip, PhotoThumbs } from "../components/Photos";

export default function VehicleScreen({
  vehicle,
  onBack,
  onOpenCategory,
  onManage,
  onUpdateOdo,
  onOpenService,
  onAddService,
  onUpdate,
}) {
  const c = useC();
  const [odoInput, setOdoInput] = useState(String(vehicle.odometer));
  const [busy, setBusy] = useState(false);
  const allLogs = [...vehicle.logs].sort((a, b) => new Date(b.date) - new Date(a.date));
  const typeLabel = TYPE_META[vehicle.type]?.label || vehicle.type;
  const subtitle = vehicle.model ? `${vehicle.model}  ·  ${typeLabel}` : typeLabel;
  const packages = getServicePackages(vehicle);

  return (
    <View style={{ flex: 1, backgroundColor: c.bg }}>
      <TopBar title={vehicle.name} subtitle={subtitle} onBack={onBack} rightLabel="Edit" onRightPress={onManage} />
      <ScrollView
        contentContainerStyle={{ paddingHorizontal: SPACE.side, paddingTop: 12, paddingBottom: 40 }}
        keyboardShouldPersistTaps="handled"
      >
        <Card style={{ paddingHorizontal: SPACE.cardPad, paddingTop: 14, paddingBottom: 16 }}>
          <Text style={[TYPE.label, { color: c.muted, marginBottom: 4 }]}>Odometer</Text>
          <View style={{ flexDirection: "row", alignItems: "baseline", gap: 8 }}>
            <TextInputOdometer
              value={odoInput}
              onChangeText={setOdoInput}
              onBlur={() => onUpdateOdo(Number(odoInput) || 0)}
              color={c.ink}
            />
            <Text style={[TYPE.body, { color: c.muted }]}>km</Text>
          </View>
        </Card>

        <View style={{ marginTop: SPACE.block }}>
          <Label style={{ marginBottom: 10 }}>Photos</Label>
          <Card style={{ padding: SPACE.cardPad }}>
            <PhotoStrip
              uris={vehicle.photos || []}
              editable
              onChange={(photos) => onUpdate({ photos })}
            />
          </Card>
        </View>

        <View style={{ marginTop: SPACE.block }}>
          <SectionHeader>Service</SectionHeader>
          <Card>
            {packages.map((pack) => (
              <CardRow
                key={pack.id}
                onPress={() => onOpenService(pack.id)}
                chevron
                dot="hidden"
                left={pack.name}
                right={`${pack.categories.length} items`}
                divider
              />
            ))}
            <CardRow
              onPress={onAddService}
              dot="hidden"
              left={<Text style={[TYPE.meta, { color: c.ink }]}>+ Add service</Text>}
              divider={false}
            />
          </Card>
        </View>

        <View style={{ marginTop: SPACE.block }}>
          <SectionHeader>Categories</SectionHeader>
          <Card>
            {vehicle.categories.map((cat, i) => {
              const status = getStatus(vehicle, cat);
              const right = status ? attentionLabel(vehicle, cat) : null;
              return (
                <CardRow
                  key={cat.id}
                  onPress={() => onOpenCategory(cat.id)}
                  dot={status ? status.level : "hidden"}
                  left={
                    <View>
                      <Text style={[TYPE.row, { color: c.ink }]}>{cat.name}</Text>
                      <Text style={[TYPE.small, { color: c.muted, marginTop: 2 }]}>
                        {status?.last
                          ? `Last  ·  ${formatDate(status.last.date)}  ·  ${formatOdo(status.last.odometer)} km`
                          : "No entries yet"}
                      </Text>
                    </View>
                  }
                  right={right}
                  chevron
                  divider={i < vehicle.categories.length - 1}
                />
              );
            })}
          </Card>
        </View>

        <View style={{ marginTop: SPACE.block }}>
          <SectionHeader>History</SectionHeader>
          {allLogs.length === 0 ? (
            <Text style={[TYPE.small, { color: c.muted, paddingVertical: SPACE.rowY }]}>
              Nothing logged yet — tap a category to add the first entry.
            </Text>
          ) : (
            <Card>
              {allLogs.map((l, i) => (
                <CardRow
                  key={l.id}
                  dot="hidden"
                  left={
                    <View>
                      <Text style={[TYPE.row, { color: c.ink }]}>{l.categoryName}</Text>
                      <Text style={[TYPE.small, { color: c.muted, marginTop: 2 }]}>
                        {l.service ? `${l.service}  ·  ` : ""}
                        {formatOdo(l.odometer)} km
                        {formatLogLine(vehicle, l) ? `  ·  ${formatLogLine(vehicle, l)}` : ""}
                        {l.note ? `  ·  ${l.note}` : ""}
                      </Text>
                      <PhotoThumbs uris={l.photos} />
                    </View>
                  }
                  right={formatDate(l.date)}
                  divider={i < allLogs.length - 1}
                />
              ))}
            </Card>
          )}
        </View>

        <View style={{ marginTop: SPACE.block }}>
          <SectionHeader>Export</SectionHeader>
          <Card>
            <CardRow
              onPress={async () => {
                if (busy) return;
                setBusy(true);
                try {
                  await exportPdf([vehicle]);
                } catch (e) {
                  Alert.alert("Export failed", e?.message || String(e));
                } finally {
                  setBusy(false);
                }
              }}
              chevron
              dot="hidden"
              left="PDF report"
              right={busy ? "…" : undefined}
              divider={false}
            />
          </Card>
        </View>
      </ScrollView>
    </View>
  );
}

function TextInputOdometer({ value, onChangeText, onBlur, color }) {
  return (
    <TextInput
      value={value}
      onChangeText={onChangeText}
      onBlur={onBlur}
      keyboardType="numeric"
      style={[TYPE.odometer, { color, padding: 0, minWidth: 120 }]}
    />
  );
}

File: lib/screens/vehicle_screen.dart
import 'package:flutter/material.dart';

import '../data/export.dart';
import '../format.dart';
import '../models.dart';
import '../theme.dart';
import '../widgets/photos.dart';
import '../widgets/ui.dart';

class VehicleScreen extends StatefulWidget {
  final Vehicle vehicle;
  final VoidCallback onBack;
  final ValueChanged<String> onOpenCategory;
  final VoidCallback onManage;
  final ValueChanged<double> onUpdateOdo;
  final ValueChanged<String> onOpenService;
  final VoidCallback onAddService;
  final ValueChanged<List<String>> onUpdatePhotos;

  const VehicleScreen({
    super.key,
    required this.vehicle,
    required this.onBack,
    required this.onOpenCategory,
    required this.onManage,
    required this.onUpdateOdo,
    required this.onOpenService,
    required this.onAddService,
    required this.onUpdatePhotos,
  });

  @override
  State<VehicleScreen> createState() => _VehicleScreenState();
}

class _VehicleScreenState extends State<VehicleScreen> {
  late TextEditingController _odoController;
  final _odoFocus = FocusNode();
  bool _busy = false;

  @override
  void initState() {
    super.initState();
    _odoController = TextEditingController(text: widget.vehicle.odometer.round().toString());
    _odoFocus.addListener(() {
      if (!_odoFocus.hasFocus) widget.onUpdateOdo(double.tryParse(_odoController.text) ?? 0);
    });
  }

  @override
  void dispose() {
    _odoController.dispose();
    _odoFocus.dispose();
    super.dispose();
  }

  Future<void> _exportPdf() async {
    if (_busy) return;
    setState(() => _busy = true);
    try {
      await exportPdf([widget.vehicle], context);
    } catch (e) {
      if (mounted) {
        await showDialog<void>(
          context: context,
          builder: (ctx) => AlertDialog(
            title: const Text('Export failed'),
            content: Text('$e'),
            actions: [TextButton(onPressed: () => Navigator.pop(ctx), child: const Text('OK'))],
          ),
        );
      }
    } finally {
      if (mounted) setState(() => _busy = false);
    }
  }

  @override
  Widget build(BuildContext context) {
    final c = AppColors.of(context);
    final vehicle = widget.vehicle;
    final allLogs = [...vehicle.logs]..sort((a, b) => DateTime.parse(b.date).compareTo(DateTime.parse(a.date)));
    final label = typeMeta[vehicle.type]?.label ?? vehicle.type;
    final subtitle = vehicle.model.isNotEmpty ? '${vehicle.model}  ·  $label' : label;
    final packages = getServicePackages(vehicle);

    return Container(
      color: c.bg,
      child: Column(
        children: [
          TopBar(title: vehicle.name, subtitle: subtitle, onBack: widget.onBack, rightLabel: 'Edit', onRightPress: widget.onManage),
          Expanded(
            child: ListView(
              padding: const EdgeInsets.fromLTRB(AppSpace.side, 12, AppSpace.side, 40),
              keyboardDismissBehavior: ScrollViewKeyboardDismissBehavior.onDrag,
              children: [
                AppCard(
                  padding: const EdgeInsets.fromLTRB(AppSpace.cardPad, 14, AppSpace.cardPad, 16),
                  children: [
                    AppText.label('Odometer', color: c.muted),
                    const SizedBox(height: 4),
                    Row(
                      crossAxisAlignment: CrossAxisAlignment.baseline,
                      textBaseline: TextBaseline.alphabetic,
                      children: [
                        SizedBox(
                          width: 140,
                          child: TextField(
                            controller: _odoController,
                            focusNode: _odoFocus,
                            keyboardType: TextInputType.numberWithOptions(decimal: true),
                            style: AppTypography.odometer.copyWith(color: c.ink),
                            decoration: const InputDecoration(isDense: true, contentPadding: EdgeInsets.zero, border: InputBorder.none),
                          ),
                        ),
                        const SizedBox(width: 8),
                        AppText.body('km', color: c.muted),
                      ],
                    ),
                  ],
                ),
                const SizedBox(height: AppSpace.block),
                const Label('Photos', margin: EdgeInsets.only(bottom: 10)),
                AppCard(
                  padding: const EdgeInsets.all(AppSpace.cardPad),
                  children: [PhotoStrip(uris: vehicle.photos, editable: true, onChange: widget.onUpdatePhotos)],
                ),
                const SizedBox(height: AppSpace.block),
                const SectionHeader('Service'),
                AppCard(
                  margin: const EdgeInsets.only(top: 14),
                  children: [
                    for (final pack in packages)
                      CardRow(
                        onPress: () => widget.onOpenService(pack.id),
                        chevron: true,
                        left: pack.name,
                        right: '${pack.categories.length} items',
                      ),
                    CardRow(onPress: widget.onAddService, left: AppText.meta('+ Add service', color: c.ink), divider: false),
                  ],
                ),
                const SizedBox(height: AppSpace.block),
                const SectionHeader('Categories'),
                AppCard(
                  margin: const EdgeInsets.only(top: 14),
                  children: [
                    for (var i = 0; i < vehicle.categories.length; i++)
                      CardRow(
                        onPress: () => widget.onOpenCategory(vehicle.categories[i].id),
                        dot: getStatus(vehicle, vehicle.categories[i])?.level == DueLevel.overdue
                            ? null
                            : getStatus(vehicle, vehicle.categories[i])?.level,
                        dotColor: getStatus(vehicle, vehicle.categories[i])?.level == DueLevel.overdue ? (vehicle.color ?? c.accent) : null,
                        chevron: true,
                        divider: i < vehicle.categories.length - 1,
                        left: _CategorySummary(vehicle: vehicle, category: vehicle.categories[i]),
                        right: getStatus(vehicle, vehicle.categories[i]) != null ? attentionLabel(vehicle, vehicle.categories[i]) : null,
                      ),
                  ],
                ),
                const SizedBox(height: AppSpace.block),
                const SectionHeader('History'),
                if (allLogs.isEmpty)
                  Padding(
                    padding: const EdgeInsets.symmetric(vertical: AppSpace.rowY),
                    child: Text('Nothing logged yet — tap a category to add the first entry.', style: AppTypography.small.copyWith(color: c.muted)),
                  )
                else
                  AppCard(
                    margin: const EdgeInsets.only(top: 14),
                    children: [
                      for (var i = 0; i < allLogs.length; i++)
                        CardRow(
                          divider: i < allLogs.length - 1,
                          left: _LogSummary(vehicle: vehicle, log: allLogs[i]),
                          right: formatDate(allLogs[i].date),
                        ),
                    ],
                  ),
                const SizedBox(height: AppSpace.block),
                const SectionHeader('Export'),
                AppCard(
                  margin: const EdgeInsets.only(top: 14),
                  children: [
                    CardRow(onPress: _exportPdf, chevron: true, left: 'PDF report', right: _busy ? '…' : null, divider: false),
                  ],
                ),
              ],
            ),
          ),
        ],
      ),
    );
  }
}

class _CategorySummary extends StatelessWidget {
  final Vehicle vehicle;
  final Category category;
  const _CategorySummary({required this.vehicle, required this.category});

  @override
  Widget build(BuildContext context) {
    final c = AppColors.of(context);
    final status = getStatus(vehicle, category);
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        AppText.row(category.name, color: c.ink),
        Padding(
          padding: const EdgeInsets.only(top: 2),
          child: AppText.small(
            status?.last != null
                ? 'Last  ·  ${formatDate(status!.last!.date)}  ·  ${formatOdo(status.last!.odometer)} km'
                : 'No entries yet',
            color: c.muted,
          ),
        ),
      ],
    );
  }
}

class _LogSummary extends StatelessWidget {
  final Vehicle vehicle;
  final LogEntry log;
  const _LogSummary({required this.vehicle, required this.log});

  @override
  Widget build(BuildContext context) {
    final c = AppColors.of(context);
    final line = formatLogLine(vehicle, log);
    final subtitle = [
      if (log.service != null) log.service,
      '${formatOdo(log.odometer)} km',
      if (line.isNotEmpty) line,
      if (log.note.isNotEmpty) log.note,
    ].join('  ·  ');
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        AppText.row(log.categoryName, color: c.ink),
        Padding(padding: const EdgeInsets.only(top: 2), child: AppText.small(subtitle, color: c.muted)),
        PhotoThumbs(uris: log.photos),
      ],
    );
  }
}
